keps/sig-storage/3314-csi-changed-block-tracking/README.md

```diff
@@ -0,0 +1,1549 @@
+the [SnapshotMetadata](#the-snapshotmetadata-service-api) gRPC service
+of the provisioner of the VolumeSnapshots to be backed up.
+
+Before using the [SnapshotMetadata](#the-snapshotmetadata-service-api) gRPC API,
+a backup application must first obtain an authentication token using the Kubernetes
+[TokenRequest API](https://kubernetes.io/docs/reference/kubernetes-api/authentication-resources/token-request-v1/)
+with the service's audience string.
+It should establish trust with the specified CA for use in gRPC calls and
+then directly make TLS gRPC calls to the
+[SnapshotMetadata](#the-snapshotmetadata-service-api) gRPC service's
+TCP endpoint.
+The audience-scoped authentication token must be passed in the metadata
+of each RPC as the value of the ``authorization`` metadata key;
+it will be used to authorize the backup application's use of the service.
+Every RPC returns a gRPC stream through which the metadata can be recovered.
+
+The advertised [SnapshotMetadata](#the-snapshotmetadata-service-api) gRPC service's
+TCP endpoint need not actually be hosted by CSI driver vendor supplied logic,
+but could be from a new community provided
+[external-snapshot-metadata sidecar](#the-external-snapshot-metadata-sidecar)
+that communicates over a private UNIX domain socket with the CSI driver's
+implementation of the [SnapshotMetadata](#the-snapshotmetadata-service-api)
+gRPC service.
+Such a CSI driver service need only focus on the generation of the metadata
+requested;
+the sidecar is responsible for validating the authentication token,
+authorizing the backup application and validating the parameters of the RPC calls.
+The sidecar forwards the RPC call to the CSI driver service over the UNIX domain socket,
+after translating Kubernetes object names into SP object names,
+and re-streams the results back to its client.
+
+The process of accessing snapshot metadata via the sidecar is illustrated in the figure below.
+Additional information is available in the [Design Details](#design-details) section.
+
+![metadata retrieval flow](./flow.drawio.svg)
+
+### User Stories
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+
+#### Full snapshot backup
+
+A backup application needs to perform a full backup on volumes of a specific
+Kubernetes application.
+
+For each volume in the application:
+1. The backup application creates a VolumeSnapshot of a PVC that needs to be
+backed up.
+2. The backup application queries the changed block tracking (CBT) service to identify
+all the allocated data blocks in the snapshot. The CBT service returns the list
+of allocated blocks.
+3. Using the VolumeSnapshot as the source, the backup application creates a new
+PVC and mounts it with `Block` VolumeMode in a pod.
```

Comment:
This actually looks like 2 topics that are quite important:

1. volume creation from a snapshot in a different namespace (cross namespace restoring/cloning)
2. volume creation with `VolumeMode = Block`, even if the original volume had `VolumeMode = Filesystem`

CSI-drivers that want to support CBT will also need to be able to do the 2nd point. I could not immediately find a reference if CSI-drivers allow that now (it is not a workflow I have seen with Ceph-CSI, we do not test that at the moment). The CSI Specification does not seem to explicitly restrict it though.

It might be useful to mention requirements for CSI-drivers somewhere.

Reply:
I think cross namespace data source is still alpha; it will be useful here. Meanwhile, it's possible to create a snapshot in the backup application namespace from a static snapshot content, based on the source volume handle of the snapshot content from the user namespace.

Snapshot volume mode conversion is [possible](https://kubernetes.io/docs/concepts/storage/volume-snapshots/#convert-volume-mode).

Reply:
Great, thanks for the pointer!